Refactor coagulation

huiwanpnnl opened the pull request:
The coagulation process in the aerosol microphysics package is refactored. Code changes are mainly in `modal_aero_coag.F90`:

- Subroutine `mam_coag_1subarea` is split into smaller subroutines; the contents not used by MAM4 are removed; many variables names are changed to (hopefully) more self-explanatory ones.
- Unused variables, data blocks, and calculations in subroutines `getcoags_wrapper_f` and `getcoags` are removed.
- The initialization of coagulation pairs in `modal_aero_amicphys:modal_aero_amicphys_init` is wrapped in a new subroutine `set_coagulation_pairs` in `modal_aero_coag.F90`; the contents not used by MAM4 are removed.

Please note that since the subroutine `mam_coag_1subarea` is significantly restructured, the comparison of `modal_aero_coag.F90` provided below by GitHub is not very helpful. To review the changes, it would be easier to simply open the new file in an editor window and read through it.

SMS test passed.

huiwanpnnl committed:
Bring in refactored mam_coag_1subarea.F90; use mam_coag_1subarea in modal_aero_amicphys.F90

SMS test PASS

huiwanpnnl committed:
coagulation: use refactored code for initialization; remove unused dummy argument

huiwanpnnl committed:
remove a few lines not used by the MODAL_AERO_4MODE_MOM configuration

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
I really like this! Thanks for removing that hard to follow do-loop. This is much easier to read!

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys.F90:
```diff
@@ -4481,6 +4475,8 @@ subroutine modal_aero_amicphys_init( imozart, species_class,n_so4_monolayers_pca
     nspec_amode, &
     numptr_amode, numptrcw_amode, sigmag_amode
 
+use modal_aero_coag, only: set_coagulation_pairs
```
(not related to this line) There is a use statement in this file at line 18:`use modal_aero_amicphys_control`. I think we should add `only` with that statement as it will make it easier to know which variable is coming from where. I know it will be a long list but I think there is value in adding that. We can add it in this PR or the next which will deal with the interfaces.

huiwanpnnl replied:
There is another comment below about `use modal_aero_amicphys_control` in `modal_aero_coag.F90`. I've addressed that one by changing to `use ... only: ...`.

This comments, in contrast, is about `modal_aero_amicphys.F90`. Since `modal_aero_amicphys_control.F90` contains pretty much all the module variables originally declared in `modal_aero_amicphys.F90`, I think in a `use ... only: ...` statement we will end up having to list everything (i.e., 100+ variables :-) ). I would suggest we take the second option mentioned above by @singhbalwinder, i.e., defer this to a future PR. Also note that in the box model, we have split those 100+ module variables into groups related to the aerosol modes, species, individual processes, and the mapping of tracer indices between `amicphys` and the host model. It might be useful to discuss (for a future PR) whether we want to port those changes back to the global model.

singhbalwinder replied:
Let's defer that to the interface PR. For the rename process, I also saw that some variables (e.g. log of `sigmag`) are redefined in amicphys for all the modes. There may be other similar calculations that we can directly use from modal_aero_data.F90 or other places in the model. This will help us bring down the number of variables in amicphys_control

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
```diff
@@ -1,2337 +1,1363 @@
+! Revision history:
+!   Richard C. Easter, 07.04.13:  Adapted from MIRAGE2 code. Some of the 
+!                                 code came from the CMAQ model v4.6.
+!   Hui Wan 2022: Refactored code: memoved non-4-mode codes; restructured subroutines.
```
memoved->removed

huiwanpnnl replied:
Corrected.

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
```diff
@@ -1,2337 +1,1363 @@
+  use cam_logfile,     only:  iulog
 
-   use modal_aero_amicphys_control
+  use modal_aero_amicphys_control
```
Adding `only` for this use statement will make it easier to know which variables are coming form this module.

huiwanpnnl replied:
Done. I also moved some additional coagulation-specific contents from `modal_aero_amicphys.F90` and `modal_aero_amicphys_control.F90` to this module.

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
```diff
@@ -1,2337 +1,1363 @@
-
-
-!EOC
+subroutine set_coagulation_pairs( big_neg_int )
```
In this case, I think we can get rid of `big_neg_int` variable and use something like the following within the subroutine:
```
integer, parameter:: huge_neg_int = -huge(huge_int)
```
and then use it like:
```
   modefrm_coagpair(:) = huge_neg_int
   modetoo_coagpair(:) = huge_neg_int
   modeend_coagpair(:) = huge_neg_int
```

huiwanpnnl replied:
I've changed `big_neg_int` from an intent(in) argument to a local parameter:

```
integer, parameter :: big_neg_int = -huge(selected_int_kind(6))
```

singhbalwinder replied:
There is a typo in my suggestion above. We can do something like this:
`integer, parameter:: huge_neg_int = -huge(huge_neg_int)`

What you used above is also fine but the above is a more typical use of `huge`.

huiwanpnnl replied:
Out of curiosity: in your suggested line of code, we have `huge_neg_int` on both the LHS and the RHS of the assignment statement. Isn't this recursive, or did I miss some language feature in Fortran?

singhbalwinder replied:
It looks a bit odd but it is not recursive. Fortran first declares the variable of a particular type (e.g. integer, real, etc.) and then uses that type to assign a value to `huge`. It is better this way as then we don't need to go look what `select_int_kind(6)` is. We just know that the variable is assigned a `huge` value of its own type.

huiwanpnnl replied:
Ah, good to know. I'll update the code following your suggestion.

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
We have been putting units in square brackets like `[s]` for all the units. Would you please change the `()` to `[]` for these units?

huiwanpnnl replied:
Yes, I'll go through the module and make the changes.

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
We have been using "src" and "dest" for origin/source and destination respectively. It would be nice to stay consistent.
```
integer :: src_mode, dest_mode
```

huiwanpnnl replied:
I can make these changes. 

Do you want the following arrays to be renamed, too? 

```
  integer, public :: modefrm_coagpair(max_coagpair)
  integer, public :: modetoo_coagpair(max_coagpair)
  integer, public :: modeend_coagpair(max_coagpair)
```

singhbalwinder replied:
Yes, I think we can rename them as `src_mode_coagpair`, `dest_mode_coagpair`  or something similar

huiwanpnnl replied:
OK. Will do!

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
We are trying to get rid of single letter variables. If you can change `a` to something meaningful or just `aa` or if it is a coefficient, `a_coef` or `coef_a`, that will be great.

huiwanpnnl replied:
Since `a` is a constant, I renamed it to `a_const`.

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
Same is the case here with variable `r`.

huiwanpnnl replied:
Changed to `r1` in [this commit](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/b13d53649eec63411da011f54c7f43184a0b907e).

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
Can we create a new subroutine for the following calculations in "Intermodal coagulation, set up for zeroeth moment"?

huiwanpnnl replied:
Done in commit [13eb6a4](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/13eb6a4ea737604b67dddeb350f6b1facb7b306c)

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
Same here, a new subroutine for third moment.

huiwanpnnl replied:
Done in [this commit](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/15badfdd0446e21d10323d34d36f00f93dc2858e).

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
The following calculations for aitken and acumulation modes are very similar. Can we create a subroutine and call it twice (for aitken and accumulation)with different inputs ?

huiwanpnnl replied:
Good point :-)  - and done in [this commit](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/060e49d6375fc68873b3ae45cf99bbb0f4199d6c)

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
Can we rename `tmpa` to something more meaningful? It is okay to use `tmp` but can add a string like `tmp_avg` or something similar?

huiwanpnnl replied:
I changed `tmpa` to `bijqnumj` in [this commit](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/0aacb82f99405c7f4207680de523988fb31ce395) and explained the meaning in the comments: it's the intermodal coagulation coefficient bij multiplied by the number mixing ratio of mode j. 

Some of the variables in these low-level subroutines are hard to give good names to, because they correspond to mathematical expressions that are specific to the numerical algorithms in use and they don't have a clear physical meaning. By using the variable name `bijqnumj` for what used to be `tmpa`, I'm trying to spell out what the quantity being calculated is, which can be useful for readers who want to check or understand the numerical algorithm. Hope this is acceptable.

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
Can we rename `tmpc` to `exp_rate` or `tmp_exp` or something similar?

huiwanpnnl replied:
Changed to `tmp_exp` in [this commit](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/0aacb82f99405c7f4207680de523988fb31ce395).

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_coag.F90:
Can we rename `tmpc` to `tmp_coag` or something similar?

huiwanpnnl replied:
In [this commit]([607f597](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/607f597a3134a0a3aafbfc3b11ed655562404c71)), I revised the comments in the subroutine to mention that the approximated mass mixing ratio equations take the form of exponential decay equations. In that context, I changed `ratesum` to `decay_const` and `tmpc` to `decay_factor`.

singhbalwinder reviewed (changes requested):
Thanks, Hui! This is very neatly written code. I really liked that way you factored out codes. It makes it a lot easier to read. I have suggested few changes. Let me know if you have any comments or questions on these changes.

huiwanpnnl committed:
Replace "use" by "use only" in modal_aero_coag.F90

Also move some coagulation-specific contents from
modal_aero_amicphys.F90 and modal_aero_amicphys_control.F90
to modal_aero_coag.F90

huiwanpnnl committed:
Declare big_neg_int as a local parameter in modal_aero_coag.F90

huiwanpnnl committed:
Use huge_neg_int suggested by Balwinder in modal_aero_coag.F90

singhbalwinder commented:
BTW, the `fast_compile_run.sh` script now works. It will make it faster to build the model. Since you are using a version without this fix, you can just remove space before `  /share/...` in the LD_LIBRARY_PATH variable in the `.env_mach_specific.sh` file (it is a hidden file) in the case directory. After that execute` bash fast_compile_run.sh` to build and submit a run.

huiwanpnnl committed:
Variable name changes: modefrm, modetoo, modeend -> src_mode, dest_mode, end_mode

Also start to use [] for units in comments. To be continued.

huiwanpnnl committed:
Finish updating notes on units in modal_aero_coag.F90

huiwanpnnl committed:
Change variable 'a' to 'a_const' in subr. getcoags

huiwanpnnl committed:
Change variable 'r' to 'r1' in subr. getcoags

huiwanpnnl committed:
new subr. 'intermodal_coag_rate_for_3rd_moment'

huiwanpnnl committed:
new subr. 'intermodal_coag_rate_for_0th_moment' and some minor reorg.

huiwanpnnl committed:
new subr. 'intramodal_coag_rate_for_0th_moment'; revise comments

huiwanpnnl committed:
cosmetic changes in subr. names and comments

huiwanpnnl committed:
Revise variable names for the update of number mixing ratio

huiwanpnnl committed:
Revise variable names for the update of mass mixing ratios

huiwanpnnl commented:
Hi @singhbalwinder, I have responded or addressed all your comments above. Please take a look and let me know if additional changes are needed.

This branch now has a minor conflict with a recent update on `refactor-maint-2.0`, which resulted in the `mam4_refactor_scripts/compare_baselines_all_tests.sh` script to give an "all-fail" result. This is because the `git merge` command in the script inserted markings of merge conflicts in `modal_aero_amicphys.F90`, which, as expected, caused `case.build` to fail. I have resolved the conflicts, pushed the updated code to a new branch, and issued a new PR (#42) to replace this one.

singhbalwinder commented:
Replaced by #42
